cita-executor/core/src/contracts/native/zk_privacy.rs

```diff
@@ -326,27 +351,32 @@ impl ZkPrivacy {
                 data.push(0u8);
             }
         }
-        let _ = ext.log(
-            vec![H256::from_str(
-                "c73b5a8f31a1a078a14123cc93687f4a59389c76caf88d5d2154d3f3ce25ff49",
-            )
-            .unwrap()],
-            &data,
-        );
+        // FixMe !!! DataProvider may offer log() function.
+        // let _ = ext.log(
+        //     vec![H256::from_str(
+        //         "c73b5a8f31a1a078a14123cc93687f4a59389c76caf88d5d2154d3f3ce25ff49",
+        //     )
+        //     .unwrap()],
+        //     &data,
+        // );
 
         trace!("send_verify OK data len {}", data.len());
-
-        Ok(GasLeft::Known(params.gas - gas_cost))
+        let gas_left = params.gas - gas_cost;
+        Ok(InterpreterResult::Normal(
+            vec![],
+            gas_left.low_u64(),
+            vec![],
+        ))
     }
 
     fn receive_verify(
         &mut self,
-        params: &ActionParams,
-        ext: &mut Ext,
-    ) -> Result<GasLeft, evm::Error> {
+        params: &ExecParams,
+        ext: &mut DataProvider,
+    ) -> Result<InterpreterResult, ExecutionError> {
         let gas_cost = U256::from(1_000_000);
         if params.gas < gas_cost {
-            return Err(evm::Error::OutOfGas);
+            return Err(ExecutionError::NativeExec("Out of gas".to_string()));
```

Comment:
It is better to define a Error Type, do not just use a `String`.

Reply:
The same as your first review advice.

Reply:
Yes.